File: app/src/main/java/com/dumuzeyn/mp3player/CardTransparencyController.java
package com.dumuzeyn.mp3player;

import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.ScrollView;
import android.widget.TextView;

final class CardTransparencyController {
    private static final int MIN_OPACITY = 35;
    private static final int MAX_OPACITY = 100;
    private final MainActivityCore host;

    CardTransparencyController(MainActivityCore host) {
        this.host = host;
    }

    String settingLabel() {
        return host.tr("Card opacity by section", "Прозрачность карточек по разделам");
    }

    void openDialog() {
        FrameLayout shade = host.uiFactory.shade();
        LinearLayout panel = host.uiFactory.panelCard();
        panel.setPadding(host.dp(16), host.dp(12), host.dp(16), host.dp(12));
        panel.addView(host.uiFactory.dialogTitle(settingLabel(), 18),
                host.uiFactory.dialogTitleParams());

        LinearLayout controls = new LinearLayout(host);
        controls.setOrientation(LinearLayout.VERTICAL);
        addControl(controls, host.tr("Songs", "Песни"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.songCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.songCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Favorites", "Избранное"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.favoriteCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.favoriteCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Playlists", "Плейлисты"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.playlistCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.playlistCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Genres", "Жанры"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.genreCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.genreCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Artists", "Исполнители"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.artistCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.artistCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Albums", "Альбомы"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.albumCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.albumCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Settings", "Настройки"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.settingsCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.settingsCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Mini-player", "Мини-плеер"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.miniPlayerCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.miniPlayerCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Application header", "Шапка приложения"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.headerCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.headerCardOpacity = value;
                    }
                });
        addControl(controls, host.tr("Dialogs", "Диалоговые окна"),
                new OpacityValue() {
                    @Override
                    public int get() {
                        return host.appearanceState.dialogCardOpacity;
                    }

                    @Override
                    public void set(int value) {
                        host.appearanceState.dialogCardOpacity = value;
                    }
                });

        ScrollView scroll = new ScrollView(host);
        scroll.addView(controls, new ScrollView.LayoutParams(-1, -2));
        panel.addView(scroll, new LinearLayout.LayoutParams(-1, 0, 1.0f));

        Button done = host.uiFactory.button(host.tr("Done", "Готово"));
        host.uiFactory.applyPrimaryButtonStyle(done);
        done.setOnClickListener(view -> {
            host.saveState();
            host.overlayHost.removeView(shade);
            host.rebuildUi();
        });
        panel.addView(done, new LinearLayout.LayoutParams(-1, host.dp(48)));
        int maxHeight = host.getResources().getDisplayMetrics().heightPixels - host.dp(96);
        shade.addView(panel, host.centerParams(host.dp(350), Math.min(host.dp(600), maxHeight)));
        host.overlayHost.addView(shade);
        host.playerUiController.updateMini();
    }

    private void addControl(LinearLayout panel, String title, OpacityValue value) {
        TextView label = host.uiFactory.text(labelText(title, value.get()), 14, true);
        panel.addView(label, new LinearLayout.LayoutParams(-1, host.dp(24)));

        SeekBar seek = new SeekBar(host);
        seek.setMax(MAX_OPACITY - MIN_OPACITY);
        seek.setProgress(value.get() - MIN_OPACITY);
        host.uiFactory.applySeekBarColors(seek);
        seek.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    value.set(MIN_OPACITY + progress);
                    label.setText(labelText(title, value.get()));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                host.saveState();
            }
        });
        panel.addView(seek, new LinearLayout.LayoutParams(-1, host.dp(32)));
    }

    private String labelText(String title, int value) {
        return title + ": " + value + "%";
    }

    private interface OpacityValue {
        int get();

        void set(int value);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/CardTransparencyController.kt
package com.dumuzeyn.mp3player

import android.widget.LinearLayout
import android.widget.FrameLayout
import android.widget.ScrollView
import android.widget.SeekBar
import kotlin.math.min

internal class CardTransparencyController(private val host: MainActivityCore) {
    fun settingLabel(): String =
        host.tr("Card opacity by section", "Прозрачность карточек по разделам")

    fun openDialog() {
        val shade = host.uiFactory.shade()
        val panel = host.uiFactory.panelCard()
        panel.setPadding(host.dp(16), host.dp(12), host.dp(16), host.dp(12))
        panel.addView(
            host.uiFactory.dialogTitle(settingLabel(), 18),
            host.uiFactory.dialogTitleParams(),
        )

        val controls = LinearLayout(host).apply { orientation = LinearLayout.VERTICAL }
        addControl(
            controls,
            host.tr("Songs", "Песни"),
            { host.appearanceState.songCardOpacity },
            { host.appearanceState.songCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Favorites", "Избранное"),
            { host.appearanceState.favoriteCardOpacity },
            { host.appearanceState.favoriteCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Playlists", "Плейлисты"),
            { host.appearanceState.playlistCardOpacity },
            { host.appearanceState.playlistCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Genres", "Жанры"),
            { host.appearanceState.genreCardOpacity },
            { host.appearanceState.genreCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Artists", "Исполнители"),
            { host.appearanceState.artistCardOpacity },
            { host.appearanceState.artistCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Albums", "Альбомы"),
            { host.appearanceState.albumCardOpacity },
            { host.appearanceState.albumCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Settings", "Настройки"),
            { host.appearanceState.settingsCardOpacity },
            { host.appearanceState.settingsCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Mini-player", "Мини-плеер"),
            { host.appearanceState.miniPlayerCardOpacity },
            { host.appearanceState.miniPlayerCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Application header", "Шапка приложения"),
            { host.appearanceState.headerCardOpacity },
            { host.appearanceState.headerCardOpacity = it },
        )
        addControl(
            controls,
            host.tr("Dialogs", "Диалоговые окна"),
            { host.appearanceState.dialogCardOpacity },
            { host.appearanceState.dialogCardOpacity = it },
        )

        val scroll = ScrollView(host)
        scroll.addView(controls, FrameLayout.LayoutParams(-1, -2))
        panel.addView(scroll, LinearLayout.LayoutParams(-1, 0, 1f))

        val done = host.uiFactory.button(host.tr("Done", "Готово"))
        host.uiFactory.applyPrimaryButtonStyle(done)
        done.setOnClickListener {
            host.saveState()
            host.overlayHost.removeView(shade)
            host.rebuildUi()
        }
        panel.addView(done, LinearLayout.LayoutParams(-1, host.dp(48)))
        val maxHeight = host.resources.displayMetrics.heightPixels - host.dp(96)
        shade.addView(panel, host.centerParams(host.dp(350), min(host.dp(600), maxHeight)))
        host.overlayHost.addView(shade)
        host.playerUiController.updateMini()
    }

    private fun addControl(
        panel: LinearLayout,
        title: String,
        getValue: () -> Int,
        setValue: (Int) -> Unit,
    ) {
        val label = host.uiFactory.text(labelText(title, getValue()), 14, true)
        panel.addView(label, LinearLayout.LayoutParams(-1, host.dp(24)))

        val seek = SeekBar(host).apply {
            max = MAX_OPACITY - MIN_OPACITY
            progress = getValue() - MIN_OPACITY
        }
        host.uiFactory.applySeekBarColors(seek)
        seek.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    setValue(MIN_OPACITY + progress)
                    label.text = labelText(title, getValue())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) = Unit

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                host.saveState()
            }
        })
        panel.addView(seek, LinearLayout.LayoutParams(-1, host.dp(32)))
    }

    private fun labelText(title: String, value: Int): String = "$title: $value%"

    companion object {
        private const val MIN_OPACITY = 35
        private const val MAX_OPACITY = 100
    }
}
